sdk/azidentity/errors.go

```diff
@@ -101,24 +101,26 @@ func (*AuthenticationFailedError) NonRetriable() {
 
 var _ errorinfo.NonRetriable = (*AuthenticationFailedError)(nil)
 
-// credentialUnavailableError indicates a credential can't attempt
-// authentication because it lacks required data or state.
-type credentialUnavailableError struct {
-	credType string
-	message  string
+// CredentialUnavailableError indicates a credential can't attempt authentication because it lacks required
+// data or state. When [ChainedTokenCredential] receives this error it will try its next credential, if any.
+type CredentialUnavailableError struct {
+	// Message describes the cause of the error
+	Message string
```

Comment:
I'm thinking this shouldn't be exported.  Is there a use-case for it to be?

Reply:
My goal was to allow user code to determine the string representation of the error, so whatever logs follow an error can be meaningful. Is there a better way to do that?

Reply:
Not sure I follow.  Do you have an example?

Reply:
https://github.com/Azure/azure-sdk-for-go/issues/19699#issuecomment-1366258523 has an example. This error may cause authentication and a client request to fail entirely, in which case a string explaining what went wrong helps in debugging. I suppose we could have an `Err error` field instead?

Reply:
Oh, is this for allowing custom credential implementations (`wrapper` in the example) to return a `CredentialUnavailableError` with a user-supplied message?  And code outside of `azidentity` doesn't actually need to `errors.As()` for this error type?  If so, it might be simpler to add a `NewCredentialUnavailableError(msg string) error` function so we can keep the error type internal.

Reply:
Although, if we do add more fields then this is easier to cope with over time.

Reply:
Hmm that's a good point, we could provide the constructor instead. Really, this type exists only to signal `ChainedTokenCredential` to try another credential. I doubt the distinction between it and any other error type is interesting in other contexts. The only `errors.As()` scenario I can think of just now is a custom chained credential implementation. I don't know why anyone would want to build their own but if they did, it would need `errors.As()` to integrate well with in-box credential types.

Reply:
A custom credential chain is an interesting point, but I agree that this seems highly unlikely.

Ultimately, it seems like what we need is a way for code external to `azidentity` be able to create an error that states "this is a non-fatal error in the context of credential chains".  I feel like a constructor is a bit cleaner in the sense that we don't have to expose the mechanism that achieves this.

Reply:
Works for me, though I think if we do end up exporting the type, we won't want the constructor.